Sources/SwiftInspectorVisitors/ClassVisitor.swift

```diff
@@ -0,0 +1,139 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import Foundation
+import SwiftSyntax
+
+public final class ClassVisitor: SyntaxVisitor {
+
+  public init(parentTypeName: String? = nil) {
+    self.parentTypeName = parentTypeName
+  }
+
+  /// All of the classes found by this visitor.
+  public var classes: [ClassInfo] {
+    [classInfo].compactMap { $0 } + innerClasses
+  }
+
+  /// Inner structs found by this visitor.
+  public private(set) var innerStructs = [StructInfo]()
+
+  // TODO: also find and nested enums
+
+  public override func visit(_ node: ClassDeclSyntax) -> SyntaxVisitorContinueKind {
+
+    guard !hasFinishedParsingClass else {
+      assertionFailure("Encountered more than one top-level class. This is a usage error: a single ClassVisitor instance should start walking only over a node of type `ClassDeclSyntax`")
+      return .skipChildren
+    }
+
+    if let classInfo = classInfo {
+      // Base case. We've previously found a class declaration, so this must be an inner class.
+      // This class visitor shouldn't recurse down into the children.
+      // Instead, we'll use a new class visitor to get the information from this class.
+      let qualifiedParentTypeName = QualifiedParentNameCreator.createNameGiven(
+        currentParentTypeName: parentTypeName,
+        currentTypeName: classInfo.name)
+
+      let innerClassVisitor = ClassVisitor(parentTypeName: qualifiedParentTypeName)
+      innerClassVisitor.walk(node)
+
+      self.innerClasses += innerClassVisitor.classes
+      // We've already gotten information from the children from our inner struct visitor.
+      return .skipChildren
+
+    } else {
+      // Recursive case. This is the first class declaration we've come across.
+      // We need to get its information and then visit children to see if there is more information we need.
+      let name = node.identifier.text
+      let typeInheritanceVisitor = TypeInheritanceVisitor()
+      typeInheritanceVisitor.walk(node)
+
+      classInfo = ClassInfo(
+        name: name,
+        inheritsFromTypes: typeInheritanceVisitor.inheritsFromTypes,
+        parentTypeName: parentTypeName)
+      return .visitChildren
+    }
+  }
+
+  public override func visitPost(_ node: ClassDeclSyntax) {
+    hasFinishedParsingClass = node.identifier.text == classInfo?.name
```

Comment:
I'm curious, should this be 
```swift
guard !hasFinishedParsingClass else { return true } 
hasFinishedParsingClass = node.identifier.text == classInfo?.name
```
?

Reply:
oooh. yeah I guess it should since the assertion won't fire in release.
```suggestion
    guard !hasFinishedParsingClass else { return true } 
    hasFinishedParsingClass = node.identifier.text == classInfo?.name
```